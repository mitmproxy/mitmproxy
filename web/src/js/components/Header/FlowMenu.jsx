import React  from "react"
import PropTypes from 'prop-types'
import { connect } from "react-redux"
import Button from "../common/Button"
import { MessageUtils } from "../../flow/utils.js"
import * as flowsActions from "../../ducks/flows"
import HideInStatic from "../common/HideInStatic";

FlowMenu.title = 'Flow'

FlowMenu.propTypes = {
    flow: PropTypes.object,
    resumeFlow: PropTypes.func.isRequired,
    killFlow: PropTypes.func.isRequired,
    replayFlow: PropTypes.func.isRequired,
    duplicateFlow: PropTypes.func.isRequired,
    removeFlow: PropTypes.func.isRequired,
    revertFlow: PropTypes.func.isRequired
}

export function FlowMenu({ flow, resumeFlow, killFlow, replayFlow, duplicateFlow, removeFlow, revertFlow }) {
    if (!flow)
        return <div/>
    return (
        <div className="flow-menu">
            <HideInStatic>
            <div className="menu-group">
                <div className="menu-content">
                    <Button title="[r]eplay flow" icon="fa-repeat text-primary"
                            onClick={() => replayFlow(flow)}>
                        Replay
                    </Button>
                    <Button title="[D]uplicate flow" icon="fa-copy text-info"
                            onClick={() => duplicateFlow(flow)}>
                        Duplicate
                    </Button>
                    <Button disabled={!flow || !flow.modified} title="revert changes to flow [V]"
                            icon="fa-history text-warning" onClick={() => revertFlow(flow)}>
                        Revert
                    </Button>
                    <Button title="[d]elete flow" icon="fa-trash text-danger"
                            onClick={() => removeFlow(flow)}>
                        Delete
                    </Button>
                </div>
                <div className="menu-legend">Flow Modification</div>
            </div>
            </HideInStatic>

            <div className="menu-group">
                <div className="menu-content">
                    <Button title="download" icon="fa-download"
                            onClick={() => window.location = MessageUtils.getContentURL(flow, flow.response)}>
                        Download
                    </Button>
                </div>
                <div className="menu-legend">Export</div>
            </div>

            <HideInStatic>
            <div className="menu-group">
                <div className="menu-content">
                    <Button disabled={!flow || !flow.intercepted} title="[a]ccept intercepted flow"
                            icon="fa-play text-success" onClick={() => resumeFlow(flow)}>
                        Resume
                    </Button>
                    <Button disabled={!flow || !flow.intercepted} title="kill intercepted flow [x]"
                            icon="fa-times text-danger" onClick={() => killFlow(flow)}>
                        Abort
                    </Button>
                </div>
                <div className="menu-legend">Interception</div>
            </div>
            </HideInStatic>
        </div>
    )
}

export default connect(
    state => ({
        flow: state.flows.byId[state.flows.selected[0]],
    }),
    {
        resumeFlow: flowsActions.resume,
        killFlow: flowsActions.kill,
        replayFlow: flowsActions.replay,
        duplicateFlow: flowsActions.duplicate,
        removeFlow: flowsActions.remove,
        revertFlow: flowsActions.revert,
    }
)(FlowMenu)
