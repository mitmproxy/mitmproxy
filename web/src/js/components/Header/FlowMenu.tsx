import * as React from "react";
import Button from "../common/Button"
import {canReplay, MessageUtils} from "../../flow/utils"
import HideInStatic from "../common/HideInStatic";
import {useAppDispatch, useAppSelector} from "../../ducks";
import * as flowActions from "../../ducks/flows"
import {
    duplicate as duplicateFlow,
    kill as killFlow,
    remove as removeFlow,
    replay as replayFlow,
    resume as resumeFlow,
    revert as revertFlow
} from "../../ducks/flows"
import Dropdown, {MenuItem} from "../common/Dropdown";
import {copy} from "../../flow/export";
import {Flow} from "../../flow";

FlowMenu.title = 'Flow'

export default function FlowMenu(): JSX.Element {
    const dispatch = useAppDispatch(),
        flow = useAppSelector(state => state.flows.byId[state.flows.selected[0]])

    if (!flow)
        return <div/>
    return (
        <div className="flow-menu">
            <HideInStatic>
                <div className="menu-group">
                    <div className="menu-content">
                        <Button title="[r]eplay flow" icon="fa-repeat text-primary"
                                onClick={() => dispatch(replayFlow(flow))}
                                disabled={!canReplay(flow)}
                        >
                            Replay
                        </Button>
                        <Button title="[D]uplicate flow" icon="fa-copy text-info"
                                onClick={() => dispatch(duplicateFlow(flow))}>
                            Duplicate
                        </Button>
                        <Button disabled={!flow || !flow.modified} title="revert changes to flow [V]"
                                icon="fa-history text-warning" onClick={() => dispatch(revertFlow(flow))}>
                            Revert
                        </Button>
                        <Button title="[d]elete flow" icon="fa-trash text-danger"
                                onClick={() => dispatch(removeFlow(flow))}>
                            Delete
                        </Button>
                        <MarkButton flow={flow}/>
                    </div>
                    <div className="menu-legend">Flow Modification</div>
                </div>
            </HideInStatic>

            <div className="menu-group">
                <div className="menu-content">
                    <DownloadButton flow={flow}/>
                    <ExportButton flow={flow}/>
                </div>
                <div className="menu-legend">Export</div>
            </div>

            <HideInStatic>
                <div className="menu-group">
                    <div className="menu-content">
                        <Button disabled={!flow || !flow.intercepted} title="[a]ccept intercepted flow"
                                icon="fa-play text-success" onClick={() => dispatch(resumeFlow(flow))}>
                            Resume
                        </Button>
                        <Button disabled={!flow || !flow.intercepted} title="kill intercepted flow [x]"
                                icon="fa-times text-danger" onClick={() => dispatch(killFlow(flow))}>
                            Abort
                        </Button>
                    </div>
                    <div className="menu-legend">Interception</div>
                </div>
            </HideInStatic>
        </div>
    )
}

// Reference: https://stackoverflow.com/a/63627688/9921431
const openInNewTab = (url) => {
  const newWindow = window.open(url, '_blank', 'noopener,noreferrer')
  if (newWindow) newWindow.opener = null
}

function DownloadButton({flow}: { flow: Flow }) {
    if (flow.type !== "http")
        return <Button icon="fa-download" onClick={() => 0} disabled>Download</Button>;

    if (flow.request.contentLength && !flow.response?.contentLength) {
        return <Button icon="fa-download"
                       onClick={() => openInNewTab(MessageUtils.getContentURL(flow, flow.request))}
        >Download</Button>
    }
    if (flow.response) {
        const response = flow.response;
        if (!flow.request.contentLength && flow.response.contentLength) {
            return <Button icon="fa-download"
                           onClick={() => openInNewTab(MessageUtils.getContentURL(flow, response))}
            >Download</Button>
        }
        if (flow.request.contentLength && flow.response.contentLength) {
            return <Dropdown text={
                <Button icon="fa-download" onClick={() => 1}>Downloadâ–¾</Button>
            } options={{"placement": "bottom-start"}}>
                <MenuItem onClick={() => openInNewTab(MessageUtils.getContentURL(flow, flow.request))}>Download
                    request</MenuItem>
                <MenuItem onClick={() => openInNewTab(MessageUtils.getContentURL(flow, response))}>Download
                    response</MenuItem>
            </Dropdown>
        }
    }

    return null;
}

function ExportButton({flow}: { flow: Flow }) {
    return <Dropdown className="" text={
        <Button title="Export flow." icon="fa-clone" onClick={() => 1}
                disabled={flow.type !== "http"}>Exportâ–¾</Button>
    } options={{"placement": "bottom-start"}}>
        <MenuItem onClick={() => copy(flow, "raw_request")}>Copy raw request</MenuItem>
        <MenuItem onClick={() => copy(flow, "raw_response")}>Copy raw response</MenuItem>
        <MenuItem onClick={() => copy(flow, "raw")}>Copy raw request and response</MenuItem>
        <MenuItem onClick={() => copy(flow, "curl")}>Copy as cURL</MenuItem>
        <MenuItem onClick={() => copy(flow, "httpie")}>Copy as HTTPie</MenuItem>
    </Dropdown>
}


const markers = {
    ":red_circle:": "ðŸ”´",
    ":orange_circle:": "ðŸŸ ",
    ":yellow_circle:": "ðŸŸ¡",
    ":green_circle:": "ðŸŸ¢",
    ":large_blue_circle:": "ðŸ”µ",
    ":purple_circle:": "ðŸŸ£",
    ":brown_circle:": "ðŸŸ¤",
}

function MarkButton({flow}: { flow: Flow }) {
    const dispatch = useAppDispatch();
    return <Dropdown className="" text={
        <Button title="mark flow" icon="fa-paint-brush text-success" onClick={() => 1}>Markâ–¾</Button>
    } options={{"placement": "bottom-start"}}>
        <MenuItem onClick={() => dispatch(flowActions.update(flow, {marked: ""}))}>âšª (no
            marker)</MenuItem>
        {Object.entries(markers).map(([name, sym]) =>
            <MenuItem
                key={name}
                onClick={() => dispatch(flowActions.update(flow, {marked: name}))}>
                {sym} {name.replace(/[:_]/g, " ")}
            </MenuItem>
        )}
    </Dropdown>
}
