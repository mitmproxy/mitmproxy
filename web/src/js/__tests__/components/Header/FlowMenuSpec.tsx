import * as React from "react";
import FlowMenu from "../../../components/Header/FlowMenu";
import { render } from "../../test-utils";

test("FlowMenu", async () => {
    const { asFragment } = render(<FlowMenu />);
    expect(asFragment()).toMatchSnapshot();
});
