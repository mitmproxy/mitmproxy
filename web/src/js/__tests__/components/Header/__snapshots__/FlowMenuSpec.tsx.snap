// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowMenu 1`] = `
<DocumentFragment>
  <div
    class="flow-menu"
  >
    <div
      class="menu-group"
    >
      <div
        class="menu-content"
      >
        <button
          class="btn btn-default"
          disabled=""
          title="[r]eplay flow"
        >
          <i
            class="fa fa-repeat text-primary"
          />
           Replay
        </button>
        <button
          class="btn btn-default"
          title="[D]uplicate flow"
        >
          <i
            class="fa fa-copy text-info"
          />
           Duplicate
        </button>
        <button
          class="btn btn-default"
          disabled=""
          title="revert changes to flow [V]"
        >
          <i
            class="fa fa-history text-warning"
          />
           Revert
        </button>
        <button
          class="btn btn-default"
          title="[d]elete flow"
        >
          <i
            class="fa fa-trash text-danger"
          />
           Delete
        </button>
        <a
          class=""
          href="#"
        >
          <button
            class="btn btn-default"
            title="mark flow"
          >
            <i
              class="fa fa-paint-brush text-success"
            />
             Mark▾
          </button>
        </a>
      </div>
      <div
        class="menu-legend"
      >
        Flow Modification
      </div>
    </div>
    <div
      class="menu-group"
    >
      <div
        class="menu-content"
      >
        <a
          class=""
          href="#"
        >
          <button
            class="btn btn-default"
          >
            <i
              class="fa fa-download"
            />
             Download▾
          </button>
        </a>
        <a
          class=""
          href="#"
        >
          <button
            class="btn btn-default"
            title="Export flow."
          >
            <i
              class="fa fa-clone"
            />
             Export▾
          </button>
        </a>
      </div>
      <div
        class="menu-legend"
      >
        Export
      </div>
    </div>
    <div
      class="menu-group"
    >
      <div
        class="menu-content"
      >
        <button
          class="btn btn-default"
          disabled=""
          title="[a]ccept intercepted flow"
        >
          <i
            class="fa fa-play text-success"
          />
           Resume
        </button>
        <button
          class="btn btn-default"
          disabled=""
          title="kill intercepted flow [x]"
        >
          <i
            class="fa fa-times text-danger"
          />
           Abort
        </button>
      </div>
      <div
        class="menu-legend"
      >
        Interception
      </div>
    </div>
  </div>
</DocumentFragment>
`;
